### PR TITLE
support containerd/cri

### PR DESCRIPTION
**Pull request by dweomer**
Adapts changes suggested by @crosbymichael:
- https://github.com/containerd/cri/pull/1246#issuecomment-629279591

Unsure if this works because #97. Any pointers to resolve the installation issue would be greatly appreciated.

### CODE REVIEW

**Comment by crosbymichael** [comment]
I remember getting those same errors when I tried to build the package.  I think I resolved it by copying the file into the systems install, or it was different for each disto version.  I can’t remember off the top of my head right now but maybe someone else has a quick suggestion on how to install.

**Comment by dweomer** [comment]
Hmm, @ibuildthecloud pointed me at https://github.com/containers/container-selinux/tree/RHEL7.5, will give this a try.

**Comment by dweomer** [comment]
BTW, thank you @crosbymichael for taking a look at this. I had another question with regards to your changes. Isn't something like the following also going to be necessary?
```
container_runtime_tmpfs_t       '/var/run/containerd/[^/]*/sandboxes/[^/]*/shm(/.*)?'
```

**Comment by dweomer** [comment]
> Hmm, @ibuildthecloud pointed me at https://github.com/containers/container-selinux/tree/RHEL7.5, will give this a try.

The RHEL7.5 branch is a no:
```
[vagrant@localhost container-selinux]$ git status
# On branch RHEL7.5
nothing to commit, working directory clean
[vagrant@localhost container-selinux]$ make install-policy
make -f /usr/share/selinux/devel/Makefile container.pp
make[1]: Entering directory `/home/vagrant/container-selinux'
/usr/share/selinux/devel/include/contrib/container.if:14: Error: duplicate definition of container_runtime_domtrans(). Original definition on 14.
/usr/share/selinux/devel/include/contrib/container.if:33: Error: duplicate definition of container_runtime_exec(). Original definition on 61.
/usr/share/selinux/devel/include/contrib/container.if:52: Error: duplicate definition of container_search_lib(). Original definition on 98.
/usr/share/selinux/devel/include/contrib/container.if:71: Error: duplicate definition of container_exec_lib(). Original definition on 117.
/usr/share/selinux/devel/include/contrib/container.if:90: Error: duplicate definition of container_read_lib_files(). Original definition on 136.
/usr/share/selinux/devel/include/contrib/container.if:109: Error: duplicate definition of container_read_share_files(). Original definition on 155.
/usr/share/selinux/devel/include/contrib/container.if:131: Error: duplicate definition of container_exec_share_files(). Original definition on 238.
/usr/share/selinux/devel/include/contrib/container.if:149: Error: duplicate definition of container_manage_lib_files(). Original definition on 275.
/usr/share/selinux/devel/include/contrib/container.if:169: Error: duplicate definition of container_manage_lib_dirs(). Original definition on 332.
/usr/share/selinux/devel/include/contrib/container.if:205: Error: duplicate definition of container_lib_filetrans(). Original definition on 368.
/usr/share/selinux/devel/include/contrib/container.if:223: Error: duplicate definition of container_read_pid_files(). Original definition on 386.
/usr/share/selinux/devel/include/contrib/container.if:242: Error: duplicate definition of container_systemctl(). Original definition on 405.
/usr/share/selinux/devel/include/contrib/container.if:267: Error: duplicate definition of container_rw_sem(). Original definition on 430.
/usr/share/selinux/devel/include/contrib/container.if:285: Error: duplicate definition of container_use_ptys(). Original definition on 467.
/usr/share/selinux/devel/include/contrib/container.if:303: Error: duplicate definition of container_filetrans_named_content(). Original definition on 485.
/usr/share/selinux/devel/include/contrib/container.if:336: Error: duplicate definition of container_stream_connect(). Original definition on 538.
/usr/share/selinux/devel/include/contrib/container.if:355: Error: duplicate definition of container_spc_stream_connect(). Original definition on 559.
/usr/share/selinux/devel/include/contrib/container.if:376: Error: duplicate definition of container_admin(). Original definition on 580.
/usr/share/selinux/devel/include/contrib/container.if:423: Error: duplicate definition of container_spc_read_state(). Original definition on 765.
/usr/share/selinux/devel/include/contrib/container.if:441: Error: duplicate definition of container_auth_domtrans(). Original definition on 627.
/usr/share/selinux/devel/include/contrib/container.if:460: Error: duplicate definition of container_auth_exec(). Original definition on 646.
/usr/share/selinux/devel/include/contrib/container.if:479: Error: duplicate definition of container_auth_stream_connect(). Original definition on 665.
/usr/share/selinux/devel/include/contrib/container.if:498: Error: duplicate definition of container_runtime_typebounds(). Original definition on 684.
container.if:14: Error: duplicate definition of container_runtime_domtrans(). Original definition on 14.
container.if:41: Error: duplicate definition of container_runtime_run(). Original definition on 41.
container.if:61: Error: duplicate definition of container_runtime_exec(). Original definition on 61.
container.if:80: Error: duplicate definition of container_read_state(). Original definition on 80.
container.if:98: Error: duplicate definition of container_search_lib(). Original definition on 98.
container.if:117: Error: duplicate definition of container_exec_lib(). Original definition on 117.
container.if:136: Error: duplicate definition of container_read_lib_files(). Original definition on 136.
container.if:155: Error: duplicate definition of container_read_share_files(). Original definition on 155.
container.if:176: Error: duplicate definition of container_runtime_read_tmpfs_files(). Original definition on 176.
container.if:197: Error: duplicate definition of container_manage_share_files(). Original definition on 197.
container.if:218: Error: duplicate definition of container_manage_share_dirs(). Original definition on 218.
container.if:238: Error: duplicate definition of container_exec_share_files(). Original definition on 238.
container.if:256: Error: duplicate definition of container_manage_config_files(). Original definition on 256.
container.if:275: Error: duplicate definition of container_manage_lib_files(). Original definition on 275.
container.if:295: Error: duplicate definition of container_manage_files(). Original definition on 295.
container.if:314: Error: duplicate definition of container_manage_dirs(). Original definition on 314.
container.if:332: Error: duplicate definition of container_manage_lib_dirs(). Original definition on 332.
container.if:368: Error: duplicate definition of container_lib_filetrans(). Original definition on 368.
container.if:386: Error: duplicate definition of container_read_pid_files(). Original definition on 386.
container.if:405: Error: duplicate definition of container_systemctl(). Original definition on 405.
container.if:430: Error: duplicate definition of container_rw_sem(). Original definition on 430.
container.if:449: Error: duplicate definition of container_append_file(). Original definition on 449.
container.if:467: Error: duplicate definition of container_use_ptys(). Original definition on 467.
container.if:485: Error: duplicate definition of container_filetrans_named_content(). Original definition on 485.
container.if:538: Error: duplicate definition of container_stream_connect(). Original definition on 538.
container.if:559: Error: duplicate definition of container_spc_stream_connect(). Original definition on 559.
container.if:580: Error: duplicate definition of container_admin(). Original definition on 580.
container.if:627: Error: duplicate definition of container_auth_domtrans(). Original definition on 627.
container.if:646: Error: duplicate definition of container_auth_exec(). Original definition on 646.
container.if:665: Error: duplicate definition of container_auth_stream_connect(). Original definition on 665.
container.if:684: Error: duplicate definition of container_runtime_typebounds(). Original definition on 684.
container.if:703: Error: duplicate definition of container_runtime_entrypoint(). Original definition on 703.
container.if:710: Error: duplicate definition of docker_exec_lib(). Original definition on 710.
container.if:714: Error: duplicate definition of docker_read_share_files(). Original definition on 714.
container.if:718: Error: duplicate definition of docker_exec_share_files(). Original definition on 718.
container.if:722: Error: duplicate definition of docker_manage_lib_files(). Original definition on 722.
container.if:727: Error: duplicate definition of docker_manage_lib_dirs(). Original definition on 727.
container.if:731: Error: duplicate definition of docker_lib_filetrans(). Original definition on 731.
container.if:735: Error: duplicate definition of docker_read_pid_files(). Original definition on 735.
container.if:739: Error: duplicate definition of docker_systemctl(). Original definition on 739.
container.if:743: Error: duplicate definition of docker_use_ptys(). Original definition on 743.
container.if:747: Error: duplicate definition of docker_stream_connect(). Original definition on 747.
container.if:751: Error: duplicate definition of docker_spc_stream_connect(). Original definition on 751.
container.if:765: Error: duplicate definition of container_spc_read_state(). Original definition on 765.
container.if:784: Error: duplicate definition of container_domain_template(). Original definition on 784.
container.if:812: Error: duplicate definition of container_spc_rw_pipes(). Original definition on 812.
Compiling targeted container module
/usr/bin/checkmodule:  loading policy configuration from tmp/container.tmp
container.te:486:ERROR 'syntax error' at token 'fs_manage_fusefs_named_sockets' on line 18289:
 	
fs_manage_fusefs_named_sockets(container_domain)
/usr/bin/checkmodule:  error(s) encountered while parsing configuration
make[1]: *** [tmp/container.mod] Error 1
make[1]: Leaving directory `/home/vagrant/container-selinux'
make: *** [container.pp] Error 2
[vagrant@localhost container-selinux]$ 
```

File: container.fc [hunk 1]
```diff
@@ -54,6 +57,9 @@
 /var/lib/docker/overlay(/.*)?	gen_context(system_u:object_r:container_ro_file_t,s0)
 /var/lib/docker/overlay2(/.*)?	gen_context(system_u:object_r:container_ro_file_t,s0)
 
+/var/lib/containerd(/.*)?	gen_context(system_u:object_r:container_var_lib_t,s0)
+/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots(/.*)?	gen_context(system_u:object_r:container_share_t,s0)
```
Comment (rhatdan) [inline container.fc:61]: Could you generalize this a little.

/var/lib/containerd/.*/snapshots(/.*)?

Also we no longer use container_share_t, please use container_ro_file_t.

**Comment by rhatdan** [comment]
Also add transition rule for containerd in container_filetrans_named_content in container.if

**Comment by dweomer** [comment]
@rhatdan I appreciate the feedback. To make this easier to test for me do you have guidance on how to make it work under centos7?

**Comment by rhatdan** [comment]
Check out the repo and then checkout the RHEL7.5 branch.

**Comment by rhatdan** [comment]
It would be best to test this policy on Fedora 32 or Centos 8 and then once we know it works backport fixes to the RHEL7.5 branch for centos7.

**Comment by dweomer** [comment]
> It would be best to test this policy on Fedora 32 or Centos 8 and then once we know it works backport fixes to the RHEL7.5 branch for centos7.

Agreed. Working on it (still have some failures running critest against containerd/cri the 1.4.0-beta but I did switch to centos8):
```
[vagrant@centos8 cri]$ sudo grep denied /var/log/audit/audit.log
type=AVC msg=audit(1594982029.763:1626): avc:  denied  { search } for  pid=40743 comm="top" name="13" dev="proc" ino=146110 scontext=system_u:system_r:container_t:s0:c575,c634 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=dir permissive=0
type=AVC msg=audit(1594982034.152:1637): avc:  denied  { open } for  pid=41942 comm="nginx" path="pipe:[149304]" dev="pipefs" ino=149304 scontext=system_u:system_r:container_t:s0:c526,c939 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982034.154:1638): avc:  denied  { open } for  pid=41942 comm="nginx" path="pipe:[149304]" dev="pipefs" ino=149304 scontext=system_u:system_r:container_t:s0:c526,c939 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982034.202:1639): avc:  denied  { open } for  pid=42019 comm="nginx" path="pipe:[150010]" dev="pipefs" ino=150010 scontext=system_u:system_r:container_t:s0:c387,c969 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982034.202:1640): avc:  denied  { open } for  pid=42019 comm="nginx" path="pipe:[150010]" dev="pipefs" ino=150010 scontext=system_u:system_r:container_t:s0:c387,c969 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982034.961:1649): avc:  denied  { open } for  pid=42540 comm="httpd" path="pipe:[151536]" dev="pipefs" ino=151536 scontext=system_u:system_r:container_t:s0:c474,c1017 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982036.639:1652): avc:  denied  { open } for  pid=43236 comm="nginx" path="pipe:[154567]" dev="pipefs" ino=154567 scontext=system_u:system_r:container_t:s0:c531,c546 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982036.642:1653): avc:  denied  { open } for  pid=43236 comm="nginx" path="pipe:[154567]" dev="pipefs" ino=154567 scontext=system_u:system_r:container_t:s0:c531,c546 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982036.743:1655): avc:  denied  { search } for  pid=43356 comm="top" name="1" dev="proc" ino=155021 scontext=system_u:system_r:container_t:s0:c4,c5 tcontext=system_u:system_r:container_t:s0:c872,c943 tclass=dir permissive=0
type=AVC msg=audit(1594982037.050:1658): avc:  denied  { open } for  pid=43587 comm="nginx" path="pipe:[155583]" dev="pipefs" ino=155583 scontext=system_u:system_r:container_t:s0:c378,c736 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982037.053:1659): avc:  denied  { open } for  pid=43587 comm="nginx" path="pipe:[155583]" dev="pipefs" ino=155583 scontext=system_u:system_r:container_t:s0:c378,c736 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982047.129:1689): avc:  denied  { search } for  pid=47108 comm="top" name="1" dev="proc" ino=169075 scontext=system_u:system_r:container_t:s0 tcontext=system_u:system_r:container_t:s0:c203,c354 tclass=dir permissive=0
type=AVC msg=audit(1594982047.173:1690): avc:  denied  { open } for  pid=47107 comm="httpd" path="pipe:[168717]" dev="pipefs" ino=168717 scontext=system_u:system_r:container_t:s0:c184,c344 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982050.799:1700): avc:  denied  { search } for  pid=48061 comm="top" name="13" dev="proc" ino=172508 scontext=system_u:system_r:container_t:s0:c115,c413 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=dir permissive=0
type=AVC msg=audit(1594982051.765:1703): avc:  denied  { open } for  pid=48325 comm="nginx" path="pipe:[173580]" dev="pipefs" ino=173580 scontext=system_u:system_r:container_t:s0:c419,c429 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982051.766:1704): avc:  denied  { open } for  pid=48325 comm="nginx" path="pipe:[173580]" dev="pipefs" ino=173580 scontext=system_u:system_r:container_t:s0:c419,c429 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982056.986:1720): avc:  denied  { open } for  pid=49942 comm="nginx" path="pipe:[179666]" dev="pipefs" ino=179666 scontext=system_u:system_r:container_t:s0:c984,c995 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982056.986:1721): avc:  denied  { open } for  pid=49942 comm="nginx" path="pipe:[179666]" dev="pipefs" ino=179666 scontext=system_u:system_r:container_t:s0:c984,c995 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982064.204:1744): avc:  denied  { open } for  pid=52433 comm="httpd" path="pipe:[188857]" dev="pipefs" ino=188857 scontext=system_u:system_r:container_t:s0:c674,c982 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982069.362:1756): avc:  denied  { search } for  pid=53832 comm="top" name="14" dev="proc" ino=195602 scontext=system_u:system_r:container_t:s0:c295,c468 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=dir permissive=0
type=AVC msg=audit(1594982074.452:1763): avc:  denied  { open } for  pid=54711 comm="nginx" path="pipe:[198342]" dev="pipefs" ino=198342 scontext=system_u:system_r:container_t:s0:c215,c717 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
type=AVC msg=audit(1594982074.453:1764): avc:  denied  { open } for  pid=54711 comm="nginx" path="pipe:[198342]" dev="pipefs" ino=198342 scontext=system_u:system_r:container_t:s0:c215,c717 tcontext=unconfined_u:system_r:container_runtime_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0
[vagrant@centos8 cri]$ 
```
The `pipefs` entries look a lot like https://bugzilla.redhat.com/show_bug.cgi?id=1456963 but that is supposed to be fixed?

**Comment by dweomer** [comment]
@rhatdan I've added the transition(s) to `container_filetrans_named_content` in `container.if` as requested. These changes resolve 10 of the 11 outstanding critest failures that I was seeing with containerd on fedora32. I am working on verifying the last but I believe that failure is on purpose and should not hold up this PR. Thank you for the assistance!

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**